### PR TITLE
`thread::spawn`, Send, Arc, Mutex

### PR DESCRIPTION
**Pull request by listochkin**

### CODE REVIEW

File: training-slides/src/deconstructing-send-arc-mutex.md [hunk 1]
```diff
@@ -0,0 +1,400 @@
+<!-- markdownlint-disable MD031 MD033 MD037 -->
+# Deconstructing Send, Arc, and Mutex
+
+## `tread::spawn` Function
```
Comment (jonathanpallant) [inline training-slides/src/deconstructing-send-arc-mutex.md:4]:
```suggestion
## `thread::spawn` Function
```

File: training-slides/src/deconstructing-send-arc-mutex.md [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+    T: Send + 'static,
+{
+    // ...
+}
+```
+
+## Quick Primer on Rust Closures
+
+* 3 categories of data
+  * data the closure *closes over* / *captures*: **Upvars**
```
Comment (jonathanpallant) [inline training-slides/src/deconstructing-send-arc-mutex.md:20]: oooh, TIL. We should call them that in the closures deck.

File: training-slides/src/deconstructing-send-arc-mutex.md [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+  * closure doesn't accept any parameters
+  * closure can *consume upvars* ("FnOnce")
+* `F: Send + 'static`
+  * applies to *upvars*
+* `T: Send + 'static`
+  * applies to returned value
+
+## `T: 'static`
+
+One of two cases:
```
Comment (jonathanpallant) [inline training-slides/src/deconstructing-send-arc-mutex.md:46]: Perhaps:

```suggestion
Two allowable options:
```

File: training-slides/src/deconstructing-send-arc-mutex.md [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+
+* the type doesn't have any references inside ("Owned data")
+  * `struct User { name: String }`
+* the references inside the type are `'static`
+  * `struct Db { connection_string: &'static str }`
+
+## Why `F: 'static` and `T: 'static`?
+
+* applies to data passed from parent thread to child thread or vise versa
```
Comment (jonathanpallant) [inline training-slides/src/deconstructing-send-arc-mutex.md:55]:
```suggestion
* applies to data passed from parent thread to child thread or vice-versa
```

File: training-slides/src/deconstructing-send-arc-mutex.md [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+* prevents passing references to local variables
+  * one thread can finish before the other and such references may become invalid
+  * `+ 'static` lets borrow checker detect these situations at compile time
```
Comment (jonathanpallant) [inline training-slides/src/deconstructing-send-arc-mutex.md:58]:
```suggestion
  * `+ 'static` avoids this by ensuring any references point to data that has the static lifetime (i.e. that lives forever)
```

File: training-slides/src/deconstructing-send-arc-mutex.md [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+
+## `T: Send`
+
+`pub unsafe auto trait Send { }`
+
+* `auto` means all types get this trait automatically
+  * opt-out instead of opt-in
+* various types in standard library implement `Send` or `!Send`
+* `unsafe` means you have to put `unsafe` keyword in front of `impl` when implementing `Send` or `!Send`
+  * precaution measure
```
Comment (jonathanpallant) [inline training-slides/src/deconstructing-send-arc-mutex.md:68]:
```suggestion
  * precautionary measure
```

File: training-slides/src/deconstructing-send-arc-mutex.md [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+
+## Why would one implement `Send` or `!Send`
+
+* Rust pointers (`*const T`, `*mut T`, `NonNull<T>`) are `!Send`
```
Comment (jonathanpallant) [inline training-slides/src/deconstructing-send-arc-mutex.md:72]: ~~Also, `UnsafeCell` is !Send, if I remember right.~~

It's Send but !Sync.

File: training-slides/src/deconstructing-send-arc-mutex.md [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+  * Use-case: what if the pointer comes from FFI library that assumes that all functions using this pointer are called from the same thread?
+* `Arc` has a `NonNull<..>` inside and becomes `!Send` automatically
+  * to override this behavior `Arc` explicitly implements `Send`
+
+## `Send` in `thread::spawn` Function
+
+`F: Send` and `T: Send` means that all data traveling from the parent thread to child thread has to be marked as `Send`
+
+* Rust compiler has to inherent knowledge of threads, but the use of marker traits and lifetime annotations let the type / borrow checker prevent data race errors
```
Comment (jonathanpallant) [inline training-slides/src/deconstructing-send-arc-mutex.md:81]:
```suggestion
* Rust compiler has no inherent knowledge of threads, but the use of marker traits and lifetime annotations let the type / borrow checker prevent data race errors
```

File: training-slides/src/deconstructing-send-arc-mutex.md [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+    data: T,
+}
+```
+
+## `Rc` vs `Arc`
+
+* `Arc` uses `AtomicUsize` for reference counting
+  * slower
+  * safe to increment / decrement from multiple threads
+* With the help of marker trait `Send` and trait bounds on `thread::spawn` Rust compiler *forces* you to use the correct type
```
Comment (jonathanpallant) [inline training-slides/src/deconstructing-send-arc-mutex.md:227]: Perhaps:

```suggestion
* With the help of marker trait `Send` and trait bounds on `thread::spawn`, the compiler *forces* you to use the correct type
```

File: training-slides/src/deconstructing-send-arc-mutex.md [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+    }
+}
+```
+<p>&nbsp<!-- run-button placeholder --></p>
+
+* `drop(guard)` also works, but extra block nicely highlights the critical section
+
+## Lessons Learned
+
+* careful use of traits and trait boundaries lets Rust compiler detect problematic multithreading code at compile time
```
Comment (jonathanpallant) [inline training-slides/src/deconstructing-send-arc-mutex.md:357]:
```suggestion
* careful use of traits and trait boundaries lets the compiler detect problematic multi-threading code at compile time
```

Or *the Rust compiler*

**Review by jonathanpallant — CHANGES_REQUESTED** [review]
This is great! Just some minor nits.

File: training-slides/src/SUMMARY.md [hunk 11]
```diff
@@ -56,6 +57,9 @@ Topics that go beyond [Applied Rust](#applied-rust).
 * [Using Types to encode State](./type-state.md)
 * [WASM](./wasm.md)
 
+## Under development
+
+* [Deconstructing Send, Arc, and Mutex](./deconstructing-send-arc-mutex.md)
```
Comment (jonathanpallant) [inline training-slides/src/SUMMARY.md:62]:
```suggestion
* [Deconstructing Send, Arc, and Mutex](./deconstructing-send-arc-mutex.md)

```

Is it under development, or is it done now?

**Review by miguelraz — APPROVED** [review]

File: training-slides/src/deconstructing-send-arc-mutex.md [hunk 12]
```diff
@@ -0,0 +1,400 @@
...
+
+Solution:
+
+* move the decision when to clean the data from compile-time to runtime
+  * use reference-counting
+
+## Attempt 1: `Rc`
+
+* `let mut log = Rc::new(vec![]);`
+* `let mut thread_log = log.clone()` now does't clone the data, but simply increases the reference count
```
Comment (jonathanpallant) [inline training-slides/src/deconstructing-send-arc-mutex.md:177]:
```suggestion
* `let mut thread_log = log.clone()` now doesn't clone the data, but simply increases the reference count
```

File: training-slides/src/deconstructing-send-arc-mutex.md [hunk 13]
```diff
@@ -0,0 +1,400 @@
...
+* `MutexGuard` in its `Drop` implementation checks if it is being dropped normally or during a `panic` unwind
+  * in later case sets a poison flag on the mutex
+* calling `lock().unwrap()` on a poisoned Mutex causes `panic`
+  * if the mutex is *"popular"* poisoning can cause many application threads to panic, too.
+* poisoning API is *problematic*
+  * `PoisonError` doesn't provide information about the panic that caused the poisoning
+  * no way to recover and revive the mutex (stays poisoned forever)
+    * `PoisonError::into_inner` *can* produce a guard even for poisoned mutexes
+
+## Critical Section "Hygiene"
```
Comment (jonathanpallant) [inline training-slides/src/deconstructing-send-arc-mutex.md:333]: This slide didn't fit on my screen :(

**Comment by jonathanpallant** [comment]
couple of typos and one slide that didn't fit (I love the Cloudflare test deployments!)

**Review by jonathanpallant — APPROVED** [review]

**Pull request merged**